Add setup.py script to build the package using setuptools

jdufresne opened the pull request:
Setuptools is a fully-featured, actively-maintained, and stable library
designed to facilitate packaging Python projects. It is the standard
library used to create Python packages by the community.

The setup.py script allows building an sdist package and a wheel
package. Installing the package with setuptools installs a gitimerge.py
module and a git-imerge command line script.

As this package is pure Python and compatible with both Python 2 & 3,
the wheel is marked as universal.

---

This continues the work started in PR #55. If work in that PR continues, I'm happy to close this one.

mhagger commented:
Thanks so much for the PR. I'm obviously no expert in setuptools (otherwise I'd have set this up myself) but it basically looks plausible to me.

But I think a couple of things are needed before this can be merged:

* Use/install instructions (e.g., in the README file). Before this PR, you could clone this repo and run `git-imerge` from the main directory without any extra steps. But since you have renamed the main file, that is no longer possible. What would be the simplest way for a user to get started using `git-imerge` if this PR is merged?

* The tests (such as they are) no longer work. Currently, you can clone the repo then run

        t/create-test-repo
        t/test-unconflicted
        t/test-conflicted
        t/test-drop

    to exercise the code. (I realize that this isn't documented.) That doesn't work anymore because of the renaming mentioned above. How do the tests have to be modified to work again?

* Does setuptools offer any help for installing the bash completion script, `git-imerge.bashcomplete`?

jdufresne commented:
Thanks for the feedback. I'm happy to work through this and then I'll post a message here when complete.

> Use/install instructions (e.g., in the README file).

Should I plan instructions to install the via PyPI? IMO, that would be most convenient to end users. If you're unfamiliar with the steps to upload to PyPI, I can help you through them.

jdufresne commented:
I've updated the PR with the following changes:

- Added installation and testing instructions to README
- Added bash completion support to the `setup.py` script through the `data_files` feature
- Added `tox` to automate testing
- Removed the `Makefile` `install` target, as now `setup.py` should be used to install the package.

Thanks! Let me know if you have any additional feedback.

mhagger commented:
Oh, another thing: it looks like `tox` spews a bunch of files under `/.tox`. That directory should probably be added to `.gitignore`.

jdufresne commented on MANIFEST.in:
```diff
@@ -0,0 +1,6 @@
+include t/create-test-repo
+include t/test-conflicted
+include t/test-drop
+include t/test-unconflicted
```
Correct.

It is customary to include test files in an sdist (source distribution).

1. Many Linux distro packagers require tests and for them to pass when adding a new package to their repository.
2. It allows end users to sanity check the state of the released package.

While these files are included in the .tar.gz file, they are not installed on the user's system.

jdufresne commented:
> So if you could give me a two-sentence reminder or a pointer to instructions, that would be helpful. I assume that git-imerge would be a suitable project name? (It's not yet claimed.)

Sure thing.

I think git-imerge would be a great name!

Once you've finalized all changes to the release, you normally do something like:

```
$ python3 setup.py sdist
$ twine upload dist/*
```

The first command builds the sdist file (source distribution). The second one uploads it. It will prompt you for a username and password. If you don't have twine installed, you can install it using `pip`:

```
$ pip install twine
```

If you're using Linux, it may also be available through your distro repository.

You can also execute twine with the command:

```
$ python3 -m twine upload dist/*
```

The following guide is an excellent source for more details:

https://python-packaging.readthedocs.io/en/latest/

jdufresne commented:
Changes:

- Cleaned up duplicate `__main__` block
- Added `.tox/` to `.gitignore`
- Included `tox.ini` in `MANIFEST.in`

mhagger commented on gitimerge.py:
```diff
@@ -4137,11 +4136,15 @@ def add_tip2_argument(subparser):
+def climain():
     try:
-        main(sys.argv[1:])
+        main()
```
It took me a while to figure out how the command-line arguments are getting through to `main`. (I wondered about that [here](https://github.com/mhagger/git-imerge/pull/143#discussion_r290060057).) It used to be that they were passed to `main` here. But now, nothing is passed to `main`, and nothing anywhere in the script references `sys.argv` anymore.

It turns out that we are relying on [this call to `parser.parse_args()`](https://github.com/mhagger/git-imerge/blob/224ce4c17d79730046b0cc2c8a61c28cf9366403/gitimerge.py#L4096), which now passes `None` to that method, which triggers `argparse`'s default behavior of taking the arguments directly from `sys.argv`.

This seems a bit magical to me. Is there a reason not to continue passing `sys.argv[1:]` explicitly here, as was done before? ("This is the standard pattern now" would an acceptable answer.) Otherwise, could `main` get a short docstring explaining that `args = None` implies that the arguments will be taken from `sys.argv`?

Otherwise, everything looks great! Thanks for your patience :smiley:

jdufresne commented:
I have restored the user of `sys.argv[1:]`.

I'm ok either way. I only removed it because it feels rather typical to me across `argparse` uses I've seen.

mhagger commented:
I just pushed a couple more commits to a branch "setuptools" under `mhagger/git-imerge`. If they look good to you, I'll merge that branch, then in a separate commit bump the version number to 1.2.0, create a tag, then push a release to PyPI. Thanks again for your patience.

jdufresne commented:
Yup, those all look good to me. Thanks!

abitrolly commented:
The ability to run `git-imerge` directly from git clone is very convenient. https://github.com/mhagger/git-imerge/tree/setuptools loses that ability.

There are also a new attempts to package Python code without `setup.py` - see https://github.com/takluyver/flit

pmav99 commented:
> I just pushed a couple more commits to a branch "setuptools" under `mhagger/git-imerge`. If they look good to you, I'll merge that branch, then in a separate commit bump the version number to 1.2.0, create a tag, then push a release to PyPI. Thanks again for your patience.

@mhagger any updates?

mhagger commented:
I don't know why I never merged this. Sorry.
